feat: Support graceful shutdown (#635)
stackabletech/kafka-operator: docs/modules/kafka/pages/usage-guide/operations/graceful-shutdown.adoc

This is equivalent to executing the `bin/kafka-server-stop.sh` command, which internally executes `kill <kafka-pid>` (https://github.com/apache/kafka/blob/2c6fb6c54472e90ae17439e62540ef3cb0426fe3/bin/kafka-server-stop.sh#L34[code]).

The broker logs the received signal as shown in the log below:

[source,text]
----
[2023-11-06 09:16:12,340] INFO Terminating process due to signal SIGTERM (org.apache.kafka.common.utils.LoggingSignalHandler)
[2023-11-06 09:16:12,341] INFO [KafkaServer id=1001] shutting down (kafka.server.KafkaServer)
[2023-11-06 09:16:12,346] INFO [KafkaServer id=1001] Starting controlled shutdown (kafka.server.KafkaServer)
[2023-11-06 09:16:12,379] INFO [Controller id=1001] Shutting down broker 1001 (kafka.controller.KafkaController)
----

=== Implementation

The https://kafka.apache.org/35/documentation/#basic_ops_restarting[Kafka documentation] does a very good job at explaining what happens during a graceful shutdown:

The Kafka cluster will automatically detect any broker shutdown or failure and elect new leaders for the partitions on that machine.

> This will occur whether a server fails or it is brought down intentionally for maintenance or configuration changes.

Reviewer: ```suggestion
This will occur if either a server fails or it is brought down intentionally for maintenance or configuration changes.
```

For the latter cases Kafka supports a more graceful mechanism for stopping a server than just killing it.
When a server is stopped gracefully it has two optimizations it will take advantage of:

> 1. It will sync all its logs to disk to avoid needing to do any log recovery when it restarts (i.e. validating the checksum for all messages in the tail of the log). Log recovery takes time so this speeds up intentional restarts.

Reviewer: ```suggestion
1. It will sync all its logs to disk to avoid the need for any log recovery when it restarts (i.e. validating the checksum for all messages in the tail of the log). Log recovery takes time, so this speeds up intentional restarts.
```

> 2. It will migrate any partitions the server is the leader for to other replicas prior to shutting down. This will make the leadership transfer faster and minimize the time each partition is unavailable to a few milliseconds.

Reviewer: ```suggestion
2. It will migrate any partitions the broker is the leader of, to other brokers prior to shutting down. This will make the leadership transfer faster and minimize the time each partition is unavailable to a few milliseconds.
```

Author: Changed to
> It will migrate any partitions the broker is the leader of, to other replicas prior to shutting down

I think there is a semantic difference between "other replica" and "other broker"
